rfcs/20211118-support-implicit-concat/implicit_concat.cpp

```diff
@@ -0,0 +1,377 @@
+    auto conv_bias2_mem = ocl_interop::make_memory(
+            conv2_pd.bias_desc(), engine, ocl_interop::memory_kind::usm);
+    auto conv_dst2_mem = ocl_interop::make_memory(
+            conv2_pd.dst_desc(), engine, ocl_interop::memory_kind::usm);
+
+    reorder(user_src2_mem, conv_src2_mem)
+            .execute(engine_stream, user_src2_mem, conv_src2_mem);
+    reorder(user_weights2_mem, conv_weights2_mem)
+            .execute(engine_stream, user_weights2_mem, conv_weights2_mem);
+    reorder(user_bias2_mem, conv_bias2_mem)
+            .execute(engine_stream, user_bias2_mem, conv_bias2_mem);
+
+    auto conv2_prim = convolution_forward(conv2_pd);
+    std::unordered_map<int, memory> conv2_args;
+    conv2_args.insert({DNNL_ARG_SRC, conv_src2_mem});
+    conv2_args.insert({DNNL_ARG_WEIGHTS, conv_weights2_mem});
+    conv2_args.insert({DNNL_ARG_BIAS, user_bias2_mem});
+    conv2_args.insert({DNNL_ARG_DST, conv_dst2_mem});
+    conv2_prim.execute(engine_stream, conv2_args);
+
+    // ========================================================================
+    // Create primitive descriptor for concat
+    // ========================================================================
+
+    // Create concat sources from convolution destinations
+    std::vector<memory::desc> concat_src_mds {
+            conv1_pd.dst_desc(), conv2_pd.dst_desc()};
+    // Let concat deduce destination for us
+    auto concat_pd = concat::primitive_desc(/*axis=*/1, concat_src_mds, engine);
+
+    auto concat_dst_md = concat_pd.dst_desc();
+    auto concat_dst_dims = concat_dst_md.dims();
+
+    // ========================================================================
+    // Validate concat destination
+    // ========================================================================
+
+    // It may happen that concat would return plain layout for destination
+    // instead of blocked, same for sources, in case descriptors are not
+    // compatible for given blocking. E.g. first input is padded.
+    //
+    // To check that, we create a memory descriptor with concat dimensions and
+    // plain `abx` layout. If they match, validation failed, explicit concat
+    // should be used.
+    auto verif_md = memory::desc(
+            concat_dst_dims, concat_dst_md.data_type(), tag::abcd);
+
+    // This is a second part of validation - that explicit and implicit concats
+    // provide the same answer.
+
+    // Execute concat for results validation
+    auto concat_dst_mem = ocl_interop::make_memory(
+            concat_dst_md, engine, ocl_interop::memory_kind::usm);
+
+    auto concat_prim = concat(concat_pd);
+    std::unordered_map<int, memory> concat_args;
+    concat_args.insert({DNNL_ARG_MULTIPLE_SRC + 0, conv_dst1_mem});
+    concat_args.insert({DNNL_ARG_MULTIPLE_SRC + 1, conv_dst2_mem});
+    concat_args.insert({DNNL_ARG_DST, concat_dst_mem});
+    concat_prim.execute(engine_stream, concat_args);
+
```

Comment:
If this is for demonstration purpose, isn't it necessary to check `conv1_dst_md` is identical to `conv2_dst_md`?

Reply:
I have added bullets 4a and 4b. Basically, if both blocked dst out of concat and submemory descs were created successfully, rest should be the library responsibility to make it run correctly (and fast (not guaranteed :) )).